include/deal.II/integrators/advection.h

```diff
@@ -311,51 +516,198 @@ namespace LocalIntegrators
                 }
+              else
+                {
+
+                  result1(i) += dx_nbeta*u2*v1;
+                  result2(i) -= dx_nbeta*u2*v2;
+                }
+            }
         }
     }
+    /**
+     * Vector: Upwind flux in the interior for weak advection operator. Matrix entries
+     * correspond to the upwind value of the trial function, multiplied by the
+     * jump of the test functions
+     * @f[
+     * a_{ij} = \int_F \left|\mathbf w
+     * \cdot \mathbf n\right|
+     * u^\uparrow
+     * (v^\uparrow-v^\downarrow)
+     * \,ds
+     * @f]
+     *
+     * The <tt>velocity</tt> is provided as a VectorSlice, having <tt>dim</tt>
+     * vectors, one for each velocity component. Each of the vectors must
+     * either have only a single entry, if t he advection velocity is
+     * constant, or have an entry for each quadrature point.
+     *
+     * The finite element can have several components, in which case each
+     * component is advected the same way.
+     */
+    template <int dim>
+    void upwind_face_residual (
+      Vector<double> &result1,
+      Vector<double> &result2,
+      const FEValuesBase<dim> &fe1,
+      const FEValuesBase<dim> &fe2,
+      const VectorSlice<const std::vector<std::vector<double> > > &input1,
+      const VectorSlice<const std::vector<std::vector<double> > > &input2,
+      const VectorSlice<const std::vector<std::vector<double> > > &velocity,
+      const double factor = 1.)
+    {
+      const unsigned int n_comp = fe1.get_fe().n_components();
+      const unsigned int n1 = fe1.dofs_per_cell;
+      AssertVectorVectorDimension(input1, n_comp, fe1.n_quadrature_points);
+      AssertVectorVectorDimension(input2, n_comp, fe2.n_quadrature_points);
+
+      // Multiply the quadrature point
+      // index below with this factor to
+      // have simpler data for constant
+      // velocities.
+      AssertDimension(velocity.size(), dim);
+      const unsigned int v_increment = (velocity[0].size() == 1) ? 0 : 1;
+      if (v_increment == 1)
+        {
+          AssertVectorVectorDimension(velocity, dim, fe1.n_quadrature_points);
+        }
+
+      for (unsigned k=0; k<fe1.n_quadrature_points; ++k)
+        {
+          double nbeta = fe1.normal_vector(k)[0] * velocity[0][k * v_increment];
+          for (unsigned int d=1; d<dim; ++d)
+            nbeta += fe1.normal_vector(k)[d] * velocity[d][k * v_increment];
+          const double dx_nbeta = factor * nbeta * fe1.JxW(k);
+
+          for (unsigned i=0; i<n1; ++i)
+            for (unsigned int d=0; d<n_comp; ++d)
+              {
+                const double v1 = fe1.shape_value_component(i,k,d);
+                const double v2 = fe2.shape_value_component(i,k,d);
+                const double u1 = input1[d][k];
+                const double u2 = input2[d][k];
+                if (nbeta >= 0)
+                  {
+                    result1(i) += dx_nbeta*v1*u1;
+                    result2(i) -= dx_nbeta*v2*u1;
```

Comment:
Sorry for nitpicking: Write dx_nbeta*u1*v1 and dx_nbeta*u1*v2, respectively, to allow the compiler to perform common subexpression elimination (it cannot do that in the other case because of associativity rules).